### PR TITLE
Error Message Grouping - TestRun Failures View

### PR DESCRIPTION
**Pull request by jeremiemarc**
Hello @pgentile ,

I've added a new view to see all failures in a testRun, grouped by their error messages.

screenshot - failures view

Jérémie

### CODE REVIEW

File: zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/domain/Scenario.java [hunk 1]
```diff
@@ -7,14 +7,7 @@
...
 import java.time.ZonedDateTime;
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Objects;
-import java.util.Optional;
-import java.util.Set;
-import java.util.UUID;
+import java.util.*;
```
Comment (pgentile) [inline zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/domain/Scenario.java]: Use real import, not wildcard imports.

Configure Intellij like this :

image

File: zucchini-ui-frontend/src/failures/components/FailuresTable.jsx [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+          </tr>
+        </thead>
+        <tbody>{rows}</tbody>
+      </Table>
+    );
+  }
+
+}
+
+FailuresTable.propTypes = {
```
Comment (pgentile) [inline zucchini-ui-frontend/src/failures/components/FailuresTable.jsx]: Integrate `propTypes` in a static field in class `FailuresTable` : there is a syntax extension by Babel that supports static properties in Javascript classes.

File: zucchini-ui-frontend/src/failures/components/FailuresTable.jsx [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    }
+    this.props.onClose();
+  };
+
+  render() {
+    const {show, errorMessage} = this.props;
+    return (
+      <Modal bsSize='large' dialogClassName='error-message-modal-dialog' show={show} onHide={this.onCloseClick}>
+        <Modal.Header closeButton>
+          <Modal.Title>{'Détails de l\'erreur'}</Modal.Title>
```
Comment (pgentile) [inline zucchini-ui-frontend/src/failures/components/FailuresTable.jsx]: You can use double quotes if the string contains a single quote.

File: zucchini-ui-frontend/src/failures/components/FailuresTable.jsx [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    });
+  };
+
+  onHideErrorMsg = () => {
+    this.setState({
+      showErrorMsgDetails: false,
+    });
+  };
+
+  handleErrorMsgRow(rowSpan, msg, scenario) {
```
Comment (pgentile) [inline zucchini-ui-frontend/src/failures/components/FailuresTable.jsx]: * Bad method name : this is not a row, this is a table cell. Better name : `createErrorMessageCell`
* You don't need scenario param, only errorMessage. Use `errorMessage` as param name, `msg` is too vague.

File: zucchini-ui-frontend/src/failures/components/FailuresTable.jsx [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    const {failedScenario, isFirstFailure, nbFailedScenarii} = this.props;
+
+    const reviewedProps = {
+      bsStyle: failedScenario.reviewed ? 'success' : 'default',
+      text: failedScenario.reviewed ? 'Oui' : 'Non',
+    };
+
+    let errorMessageRow;
+    if (isFirstFailure) {
+      errorMessageRow = this.handleErrorMsgRow(nbFailedScenarii, failedScenario.errorMessage.substring(0, 300), failedScenario);
```
Comment (pgentile) [inline zucchini-ui-frontend/src/failures/components/FailuresTable.jsx]: * The substring should be in the `handleErrorMsgRow` method.
* You should use an ellipsis (...) to indicate that the error message has been truncated. Lodash lib is available in Zucchini and has a `truncate` function : https://lodash.com/docs/4.17.4#truncate, you can use that.

File: zucchini-ui-frontend/src/failures/components/FailuresTable.jsx [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+        <FailuresDetailsDialog errorMessage={scenario.errorMessage} show={this.state.showErrorMsgDetails} onClose={this.onHideErrorMsg}/>
+      </td>
+    );
+  }
+
+  render() {
+    const {failedScenario, isFirstFailure, nbFailedScenarii} = this.props;
+
+    const reviewedProps = {
+      bsStyle: failedScenario.reviewed ? 'success' : 'default',
```
Comment (pgentile) [inline zucchini-ui-frontend/src/failures/components/FailuresTable.jsx]: You can add ternary expression directly in JSX, you don't need an extra variable.

File: zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/domain/Scenario.java (hunk 1 above)
Comment (jeremiemarc) [inline zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/domain/Scenario.java]: Review taken into account

File: zucchini-ui-frontend/src/failures/components/FailuresTable.jsx [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        <tbody>{rows}</tbody>
+      </Table>
+    );
+  }
+
+}
+
+FailuresTable.propTypes = {
+  testRunId: PropTypes.string.isRequired,
+  failures: PropTypes.object.isRequired,
```
Comment (pgentile) [inline zucchini-ui-frontend/src/failures/components/FailuresTable.jsx]: `failures` prop type must be an array, not an object.

File: zucchini-ui-frontend/src/failures/components/FailuresTable.jsx [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+import Table from 'react-bootstrap/lib/Table';
+import Label from 'react-bootstrap/lib/Label';
+import Modal from 'react-bootstrap/lib/Modal';
+import Status from '../../ui/components/Status';
+
+export default class FailuresTable extends React.Component {
+
+  render() {
+    const {failures} = this.props;
+    const rows = failures.failures.map(groupedFailures => {
```
Comment (pgentile) [inline zucchini-ui-frontend/src/failures/components/FailuresTable.jsx]: I think you should inject in your component only the `failures` array, not the full Redux store part.

File: zucchini-ui-frontend/src/failures/components/FailuresTableContainer.js [hunk 9]
```diff
@@ -0,0 +1,21 @@
+import {connect} from 'react-redux';
+import {createSelector, createStructuredSelector} from 'reselect';
+
+import FailuresTable from './FailuresTable';
+
+const selectFailures = createSelector(
```
Comment (pgentile) [inline zucchini-ui-frontend/src/failures/components/FailuresTableContainer.js]: Selector too vague. You should return `failules.failures` from the Redux store. Your selector should look like :

```
const selectFailures = createSelector(
  state => state.failures.failures,
  failures => failures,
);
```

File: zucchini-ui-frontend/src/main.less [hunk 10]
```diff
@@ -147,3 +147,8 @@
   transition: opacity linear 100ms;
   opacity: 0;
 }
+
+.error-message-modal-dialog {
+  max-width: 100%;
+  display: table;
```
Comment (pgentile) [inline zucchini-ui-frontend/src/main.less]: Why ?

File: zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/rest/ScenarioResource.java [hunk 11]
```diff
@@ -107,6 +89,21 @@ public ScenarioStats getStats(@BeanParam final GetScenariiRequestParams requestP
...
+        Consumer<ScenarioQuery> query = q -> {
+            if (!Strings.isNullOrEmpty(requestParams.getTestRunId())) {
+                q.withTestRunId(requestParams.getTestRunId());
+            }
+            if (!Strings.isNullOrEmpty(requestParams.getFeatureId())) {
+                q.withFeatureId(requestParams.getFeatureId());
+            }
+            q.havingErrorMessage();
+        };
+        return ErrorMessageGroupingUtils.computeDistance(scenarioViewAccess.getFailedScenarii(query));
```
Comment (jeremiemarc) [inline zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/rest/ScenarioResource.java]: Refactored code according to your review comment ;-)

File: zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/views/ErrorMessageGroupingUtils.java [hunk 12]
```diff
@@ -28,4 +31,30 @@ public static boolean isSimilar(String referenceText, String targetText) {
         return StringUtils.getLevenshteinDistance(referenceText, targetText, Math.round(MAX_DISTANCE_DIFF_RATIO * referenceText.length())) != -1;
     }
 
+    /**
+     * Method computing the distances between error messages outputs of failed scenarii
+     *
+     * @param scenarii All scenarii to consider during the computation
+     * @return List of failed scenarii grouped by their errorMessage
+     */
+    public static List<GroupedFailuresListItemView> computeDistance(List<FailedScenarioListItemView> scenarii) {
```
Comment (jeremiemarc) [inline zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/views/ErrorMessageGroupingUtils.java]: I moved the code directly to the `ScenarioViewAccess ` class

File: zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/views/GroupedFailuresListItemView.java [hunk 13]
```diff
@@ -0,0 +1,27 @@
+package io.zucchiniui.backend.scenario.views;
+
+import java.util.Comparator;
+import java.util.Set;
+import java.util.TreeSet;
+
+public class GroupedFailuresListItemView {
+
+    private String errorMessage;
+    private Set<FailedScenarioListItemView> failedScenarii = new TreeSet<>(Comparator.comparing(failure -> failure.getInfo().getName()));
```
Comment (jeremiemarc) [inline zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/views/GroupedFailuresListItemView.java]: OK

File: zucchini-ui-frontend/src/main.less (hunk 10 above)
Comment (jeremiemarc) [inline zucchini-ui-frontend/src/main.less]: I did not find an easier to display the full errorMessage content (without scrollbars, ...)

File: zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/views/ScenarioViewAccess.java [hunk 14]
```diff
@@ -73,6 +75,31 @@ public ScenarioViewAccess(final ScenarioDAO scenarioDAO, final TestRunRepository
...
+                GroupedFailuresListItemView noMatch = new GroupedFailuresListItemView();
+                noMatch.setErrorMessage(scenario.getErrorMessage());
+                noMatch.setFailedScenarii(new TreeSet<>(Comparator.comparing(failure -> failure.getInfo().getName())));
+                noMatch.addFailedScenario(scenario);
+                groupedFailures.add(noMatch);
+            }
+        });
+        return groupedFailures
+            .stream()
+            .sorted(Collections.reverseOrder(Comparator.comparingInt(grp -> grp.getFailedScenarii().size())))
```
Comment (pgentile) [inline zucchini-ui-backend/src/main/java/io/zucchiniui/backend/scenario/views/ScenarioViewAccess.java]: Simpler :

`Collections.reverseOrder(Comparator.comparingInt(grp -> grp.getFailedScenarii().size()))`

To :

`Comparator.comparingInt(grp -> grp.getFailedScenarii().size()).reversed()`

**Comment by pgentile** [comment]
Check last comment, add an entry in the `CHANGELOG.md` file, and I merge the PR 👍

**Pull request merged**